### PR TITLE
feat(comp/otelcol): component to filter OTEL telemetry

### LINKED ISSUE #4628: Flow: Add OpenTelemetry Collector components - Filter Processor
### Request

Please consider adding the https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/filterprocessor into Grafana Agent Flow so that we can filter spans from traces https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/filterprocessor/README.md#filter-spans-from-traces ?

### Use case

Filter spans from traces and easily migrate from Otel Collector.

### PR DESCRIPTION
**Pull request by hainenber**
#### Which issue(s) this PR fixes

Closes #4628 

#### PR Checklist

- [x] CHANGELOG.md updated
- [x] Documentation added
- [x] Tests updated
- [x] Config converters updated

### CODE REVIEW

File: docs/sources/flow/reference/components/otelcol.processor.filter.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+[OTTL boolean expressions]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/v0.85.0/pkg/ottl#boolean-expressions
+[OTTL resource context]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.85.0/pkg/ottl/contexts/ottlresource/README.md
+[OTTL scope context]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.85.0/pkg/ottl/contexts/ottlscope/README.md
+[OTTL span context]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.85.0/pkg/ottl/contexts/ottlspan/README.md
+[OTTL spanevent context]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.85.0/pkg/ottl/contexts/ottlspanevent/README.md
+[OTTL metric context]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.85.0/pkg/ottl/contexts/ottlmetric/README.md
+[OTTL datapoint context]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.85.0/pkg/ottl/contexts/ottldatapoint/README.md
+[OTTL log context]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.85.0/pkg/ottl/contexts/ottllog/README.md
```
Comment (ptodev) [inline docs/sources/flow/reference/components/otelcol.processor.filter.md:260]: Are all of those URLs used in this doc page?

File: component/otelcol/processor/filter/filter.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+type traceConfig struct {
+	Span      []string `river:"span,attr,optional"`
+	SpanEvent []string `river:"spanevent,attr,optional"`
+}
+
+type metricConfig struct {
+	Metric    []string `river:"metric,attr,optional"`
+	Datapoint []string `river:"datapoint,attr,optional"`
+}
+type logConfig struct {
```
Comment (ptodev) [inline component/otelcol/processor/filter/filter.go:38]: Would you be able to also add the `include` and `exclude` properties please?

File: component/otelcol/processor/filter/filter.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+		},
+	})
+}
+
+type traceConfig struct {
+	Span      []string `river:"span,attr,optional"`
+	SpanEvent []string `river:"spanevent,attr,optional"`
+}
+
+type metricConfig struct {
```
Comment (ptodev) [inline component/otelcol/processor/filter/filter.go:34]: Would you be able to also add the other config properties please:
* `include`
*  `exclude`
*  `regexp`

File: docs/sources/flow/reference/components/otelcol.processor.filter.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+{{% /admonition %}}
+
+{{% admonition type="note" %}}
+`otelcol.processor.filter` is a wrapper over the upstream
+OpenTelemetry Collector `filter` processor. If necessary, bug reports or feature requests
+will be redirected to the upstream repository.
+{{% /admonition %}}
+
+You can specify multiple `otelcol.processor.filter` components by giving them different labels.
+
```
Comment (ptodev) [inline docs/sources/flow/reference/components/otelcol.processor.filter.md:55]:
```suggestion

{{% admonition type="warning" %}}
Exercise caution when using `otelcol.processor.filter`:

- Make sure you understand the look of the incoming data and test the configuration thoroughly. 
  In general, use as specific a configuration as possible to lower the risk of the wrong data being dropped.
- [Orphaned Telemetry][]: The processor allows dropping spans. Dropping a span may lead to 
  orphaned spans if the dropped span is a parent. Dropping a span may lead to orphaned logs 
  if the log references the dropped span.

[Orphaned Telemetry]: https://github.com/open-telemetry/opentelemetry-collector/blob/v0.85.0/docs/standard-warnings.md#orphaned-telemetry
{{% /admonition %}}

```

**Review by ptodev — CHANGES_REQUESTED** [review]
Thank you very much for this contribution! This has been a much-requested feature. I added a few comments for things we will need to change prior to merging.

File: component/otelcol/processor/filter/filter.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+		NeedsServices: []string{otel_service.ServiceName},
+
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			fact := filterprocessor.NewFactory()
+			return processor.New(opts, fact, args.(Arguments))
+		},
+	})
+}
+
+type traceConfig struct {
```
[the remark this answers is not in the document]
Comment (hainenber) [reply component/otelcol/processor/filter/filter.go]: Thanks for the review! Done in latest commit.

File: docs/sources/flow/reference/components/otelcol.processor.filter.md (hunk 1 above)
Comment (hainenber) [reply docs/sources/flow/reference/components/otelcol.processor.filter.md]: Thanks for the review! I've removed the unused Markdown refs in latest commit.

File: docs/sources/flow/reference/components/otelcol.processor.filter.md [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+canonical: https://grafana.com/docs/agent/latest/flow/reference/components/otelcol.processor.filter/
+labels:
+  stage: experimental
+title: otelcol.processor.filter
+description: Learn about otelcol.processor.filter
+---
+
+# otelcol.processor.filter
+
+{{< docs/shared lookup="flow/stability/experimental.md" source="agent" version="<AGENT VERSION>" >}}
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.filter.md:15]:
```suggestion
{{< docs/shared lookup="flow/stability/experimental.md" source="agent" version="<AGENT_VERSION>" >}}
```

We are migrating to use an underscore in the `<AGENT_VERSION>` variable

File: docs/sources/flow/reference/components/otelcol.processor.filter.md [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+
+`otelcol.processor.filter` accepts and filters telemetry data from other `otelcol`
+components using the [OpenTelemetry Transformation Language (OTTL)][OTTL].
+OTTL statements consist of [OTTL Converter functions][], which act on paths.
+A path is a reference to a telemetry data such as:
+* Resource attributes.
+* Instrumentation scope name.
+* Span attributes.
+If any condition is met, the telemetry data is dropped (each condition is ORed together)
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.filter.md:24]:
```suggestion
If any condition is met, the telemetry data is dropped (each condition is ORed together).
```

File: docs/sources/flow/reference/components/otelcol.processor.filter.md [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+
+In addition to the [standard OTTL Converter functions][OTTL Converter functions], 
+there is also a set of metrics-only functions exclusively by the processor
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.filter.md:27]:
```suggestion
In addition to the [standard OTTL Converter functions][OTTL Converter functions], 
the following metrics-only functions are used exclusively by the processor:
```

I'm not sure this rephrase is accurate. The original looks like it's missing some words... and this is my best guess at what was intended.

File: docs/sources/flow/reference/components/otelcol.processor.filter.md [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+OpenTelemetry Collector `filter` processor. If necessary, bug reports or feature requests
+will be redirected to the upstream repository.
+{{% /admonition %}}
+
+You can specify multiple `otelcol.processor.filter` components by giving them different labels.
+
+{{% admonition type="warning" %}}
+Exercise caution when using `otelcol.processor.filter`:
+
+- Make sure you understand the look of the incoming data and test the configuration thoroughly.
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.filter.md:59]: What is meant by "understand the look"?  The meaning isn't clear and I'm not sure what to suggest here.

File: docs/sources/flow/reference/components/otelcol.processor.filter.md [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+will be redirected to the upstream repository.
+{{% /admonition %}}
+
+You can specify multiple `otelcol.processor.filter` components by giving them different labels.
+
+{{% admonition type="warning" %}}
+Exercise caution when using `otelcol.processor.filter`:
+
+- Make sure you understand the look of the incoming data and test the configuration thoroughly. 
+  In general, use as specific a configuration as possible to lower the risk of the wrong data being dropped.
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.filter.md:60]:
```suggestion
  In general, use a configuration that is as specific as possible ensure you retain only the data you want to keep.
```

I've tried to rephrase this in a more active and positive way. Does this stay true to the intent of the concept described?

File: docs/sources/flow/reference/components/otelcol.processor.filter.md [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+
+[traces]: #traces-block
+[metrics]: #metrics-block
+[logs]: #logs-block
+[output]: #output-block
+
+
+### traces block
+
+The `traces` block specifies statements which filter trace telemetry signals.
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.filter.md:112]:
```suggestion
The `traces` block specifies statements that filter trace telemetry signals. 
```

File: docs/sources/flow/reference/components/otelcol.processor.filter.md [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+The syntax of OTTL statements depends on the OTTL context. See the OpenTelemetry 
+documentation for more information:
+* [OTTL metric context][]
+* [OTTL datapoint context][]
+
+Statements are checked in order from "high level" to "low level" telemetry, in this order:
+1. `metric`
+2. `datapoint`
+
+Hence, if at least one `metric` condition is satisfied, the `datapoint` conditions will not be checked.
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.filter.md:153]:
```suggestion
If at least one `metric` condition is satisfied, the `datapoint` conditions will not be checked.
```

File: docs/sources/flow/reference/components/otelcol.processor.filter.md [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+2. `spanevent`
+
+Hence, if at least one `span` condition is satisfied, the `spanevent` conditions will not be checked.
+Only one of the statements inside the list of statements has to be satisfied.
+
+If all span events for a span are dropped, the span will be left intact.
+
+### metrics block
+
+The `metrics` block specifies statements which filter metric telemetry signals.
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.filter.md:136]:
```suggestion
The `metrics` block specifies statements that filter metric telemetry signals. 
```

File: docs/sources/flow/reference/components/otelcol.processor.filter.md [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+The syntax of OTTL statements depends on the OTTL context. See the OpenTelemetry 
+documentation for more information:
+* [OTTL span context][]
+* [OTTL spanevent context][]
+
+Statements are checked in order from "high level" to "low level" telemetry, in this order:
+1. `span`
+2. `spanevent`
+
+Hence, if at least one `span` condition is satisfied, the `spanevent` conditions will not be checked.
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.filter.md:129]:
```suggestion
If at least one `span` condition is satisfied, the `spanevent` conditions will not be checked.
```

File: docs/sources/flow/reference/components/otelcol.processor.filter.md [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+2. `datapoint`
+
+Hence, if at least one `metric` condition is satisfied, the `datapoint` conditions will not be checked.
+Only one of the statements inside the list of statements has to be satisfied.
+
+If all datapoints for a metric are dropped, the metric will also be dropped.
+
+### logs block
+
+The `logs` block specifies statements which filter log telemetry signals.
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.filter.md:160]:
```suggestion
The `logs` block specifies statements that filter log telemetry signals. 
```

**Review by clayton-cornell — COMMENTED** [review]
Some doc suggestions and questions

**Comment by ptodev** [comment]
> Appreciate your review effort @ptodev! Other comments totally makes sense to me but there are three with similar concept so I'd like to discuss jointly.
> 
> I saw that the config schema with `include`, `exclude` and alike are at brink of being deprecated, as dictated in the [ref](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/filterprocessor/README.md#alternative-config-options). Should we also implement it for a brand new Flow component? I feel like it could be technical baggage and we'll revisit for the seemingly inevitable removal.

@hainenber Thank you for pointing this out! I did not realise it. In that case I agree that we can just have the OTTL parameters. We don't need to include these other ones which will be deprecated.

File: component/otelcol/processor/filter/filter.go (hunk 3 above)
Comment (ptodev) [reply component/otelcol/processor/filter/filter.go]: As we [agreed](https://github.com/grafana/agent/pull/5562#issuecomment-1782042834), there is no need to include these.

File: component/otelcol/processor/filter/filter.go (hunk 2 above)
Comment (ptodev) [reply component/otelcol/processor/filter/filter.go]: As we [agreed](https://github.com/grafana/agent/pull/5562#issuecomment-1782042834), there is no need to include these.

File: component/otelcol/processor/filter/filter.go [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+
+type metricConfig struct {
+	Metric    []string `river:"metric,attr,optional"`
+	Datapoint []string `river:"datapoint,attr,optional"`
+}
+type logConfig struct {
+	LogRecord []string `river:"log_record,attr,optional"`
+}
+
+type Arguments struct {
```
[the remark this answers is not in the document]
Comment (ptodev) [reply component/otelcol/processor/filter/filter.go:29]: As we [agreed](https://github.com/grafana/agent/pull/5562#issuecomment-1782042834), there is no need to include these.

**Comment by ptodev** [comment]
@hainenber Would you mind addressing @clayton-cornell's comments please, and rebasing the branch? I will then take a final look and will likely merge it. Thanks again very much for your work on this component!

File: docs/sources/flow/reference/components/otelcol.processor.filter.md (hunk 10 above)
Comment (hainenber) [reply docs/sources/flow/reference/components/otelcol.processor.filter.md:60]: Very much so. Suggested change accepted. Thanks!

File: docs/sources/flow/reference/components/otelcol.processor.filter.md (hunk 9 above)
Comment (hainenber) [reply docs/sources/flow/reference/components/otelcol.processor.filter.md:59]: It should be something akin to understanding the data's schema/format. Let me wordsmith this one

**Comment by hainenber** [comment]
> @hainenber Would you mind addressing @clayton-cornell's comments please, and rebasing the branch? I will then take a final look and will likely merge it. Thanks again very much for your work on this component!

@ptodev I've addressed Clayton's comments and rebased the branch. PTAL when you're back. Thanks!

File: docs/sources/flow/reference/components/otelcol.processor.filter.md [hunk 17]
```diff
@@ -0,0 +1,302 @@
...
+  * `end_time_unix_nano - start_time_unix_nano`
+  * `sum([1, 2, 3, 4]) + (10 / 1) - 1`
+
+{{% admonition type="note" %}}
+Some characters inside River strings [need to be escaped][river-strings] with a `\` character.
+For example, the OTTL statement `attributes["grpc"] == true` 
+is written in River as `attributes[\"grpc\"] == true`.
+
+[river-strings]: {{< relref "../../config-language/expressions/types_and_values.md/#strings" >}}
+{{% /admonition %}}
```
Comment (ptodev) [inline docs/sources/flow/reference/components/otelcol.processor.filter.md:46]: The next release of the agent will support raw strings, so I updated some of the documentation in #5636. I'm happy to update the `otelcol.processor.filter` docs after your PR is merged though. There's no need to mention raw strings in your PR.

**Review by ptodev — APPROVED** [review]
Thank you, it looks good! I left a few minor comments, but the only thing which really needs to be fixed is just the non-working unit test.

**Comment by hainenber** [comment]
Thank you for another review! I've accepted most of your suggested changes and moved the `convert()` to `types.go` file. PTAL if it's OK by now.

**Review by ptodev — APPROVED** [review]

**Pull request merged**